File: sdks/js/src/span.ts
import type { SpanBarn } from './client.js'
import { generateSpanId } from './ids.js'
import type { SpanAttributes, SpanData, SpanOptions } from './types.js'

function nowMicroseconds(): number {
  return Math.floor(performance.now() * 1000) + Math.floor(Date.now() * 1000)
}

export class Span {
  private data: SpanData
  private ended: boolean = false
  private client: SpanBarn

  constructor(client: SpanBarn, name: string, traceId: string, options?: SpanOptions) {
    this.client = client
    this.data = {
      traceId,
      spanId: generateSpanId(),
      parentSpanId: options?.parentSpanId,
      name,
      service: client.getService(),
      kind: options?.kind ?? 'internal',
      status: 'unset',
      startTime: nowMicroseconds(),
      duration: 0,
      attributes: { ...options?.attributes },
      events: [],
    }
  }

  /** Get the span ID */
  getSpanId(): string {
    return this.data.spanId
  }

  /** Get the trace ID */
  getTraceId(): string {
    return this.data.traceId
  }

  /** Set a single attribute */
  setAttribute(key: string, value: string | number | boolean): this {
    if (!this.ended) {
      this.data.attributes[key] = value
    }
    return this
  }

  /** Set multiple attributes */
  setAttributes(attrs: SpanAttributes): this {
    if (!this.ended) {
      Object.assign(this.data.attributes, attrs)
    }
    return this
  }

  /** Set span status */
  setStatus(status: 'ok' | 'error'): this {
    if (!this.ended) {
      this.data.status = status
    }
    return this
  }

  /** Add an event to the span */
  addEvent(name: string, attributes?: SpanAttributes): this {
    if (!this.ended) {
      this.data.events.push({
        name,
        time: nowMicroseconds(),
        attributes,
      })
    }
    return this
  }

  /** Shorthand for setStatus('ok') */
  ok(): this {
    return this.setStatus('ok')
  }

  /** Set error status and add error attributes */
  error(err?: Error | string): this {
    this.setStatus('error')
    if (!this.ended) {
      if (err instanceof Error) {
        this.data.attributes['error.type'] = err.name
        this.data.attributes['error.message'] = err.message
        if (err.stack) {
          this.data.attributes['error.stack'] = err.stack
        }
      } else if (typeof err === 'string') {
        this.data.attributes['error.message'] = err
      }
    }
    return this
  }

  /** End the span, calculating duration and enqueueing */
  end(): void {
    if (this.ended) return
    this.ended = true
    this.data.duration = nowMicroseconds() - this.data.startTime
    this.client.enqueue(this.data)
  }

  /** Check if the span has ended (for testing) */
  isEnded(): boolean {
    return this.ended
  }
}

File: sdks/go/span.go
package spanbarn

import (
	"sync"
	"time"
)

// Span represents an in-flight operation being traced.
type Span struct {
	client *Client
	data   SpanData
	ended  bool
	mu     sync.Mutex
}

// SetAttribute sets a single attribute on the span.
func (s *Span) SetAttribute(key string, value interface{}) *Span {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return s
	}
	if s.data.Attributes == nil {
		s.data.Attributes = make(map[string]interface{})
	}
	s.data.Attributes[key] = value
	return s
}

// SetAttributes sets multiple attributes on the span.
func (s *Span) SetAttributes(attrs map[string]interface{}) *Span {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return s
	}
	if s.data.Attributes == nil {
		s.data.Attributes = make(map[string]interface{})
	}
	for k, v := range attrs {
		s.data.Attributes[k] = v
	}
	return s
}

// SetStatus sets the span status string.
func (s *Span) SetStatus(status string) *Span {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return s
	}
	s.data.Status = status
	return s
}

// AddEvent adds a timestamped event to the span.
func (s *Span) AddEvent(name string, attrs ...map[string]interface{}) *Span {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return s
	}
	ev := SpanEvent{
		Name: name,
		Time: time.Now().UnixMicro(),
	}
	if len(attrs) > 0 && attrs[0] != nil {
		ev.Attributes = attrs[0]
	}
	s.data.Events = append(s.data.Events, ev)
	return s
}

// Ok sets the span status to "ok".
func (s *Span) Ok() *Span {
	return s.SetStatus("ok")
}

// Error sets the span status to "error" and records the error message.
func (s *Span) Error(err error) *Span {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return s
	}
	s.data.Status = "error"
	if s.data.Attributes == nil {
		s.data.Attributes = make(map[string]interface{})
	}
	if err != nil {
		s.data.Attributes["error.message"] = err.Error()
	}
	return s
}

// End completes the span, calculates its duration, and enqueues it for export.
// Calling End a second time is a no-op.
func (s *Span) End() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return
	}
	s.ended = true
	s.data.Duration = time.Now().UnixMicro() - s.data.StartTime
	if s.client != nil {
		data := s.data // copy
		s.client.enqueue(&data)
	}
}
